PUBSTWO-1690 (Make transformed XML pubs available from pubs-services):

orosz-usgs opened the pull request:
Added new end point /pubs-services/publication/full/{indexId}
  New Utils class XmlUtils
  Added PUBLICATION_XML field to LinkType
  Added Publication XML link to test publication
  Updated stream test data to reflect the additional link in the test publication
  New tests: XmlUtilsTest and PwPublicationMvcServiceXmIT
  Updated tests to reflect the additional publication link.

Commit:
PUBSTWO-1690 (Make transformed XML pubs available from pubs-services):
  Added new end point /pubs-services/publication/full/{indexId}
  New Utils class XmlUtils
  Added PUBLICATION_XML field to LinkType
  Added Publication XML link to test publication
  Updated stream test data to reflect the additional link in the test publication
  New tests: XmlUtilsTest and PwPublicationMvcServiceXmIT
  Updated tests to reflect the additional publication link.

ssoper-usgs commented on src/main/java/gov/usgs/cida/pubs/utility/XmlUtils.java:
```diff
@@ -0,0 +1,188 @@
+		TransformerFactory tfactory = TransformerFactory.newInstance();
+
+		// Create a transformer for the stylesheet.
+		InputStream xslIS = new BufferedInputStream(new FileInputStream(xlsStylesheet));
```
`xlsStylesheet`  should maybe be `xslStylesheet`

dsteinich commented on src/main/java/gov/usgs/cida/pubs/utility/XmlUtils.java:
```diff
@@ -0,0 +1,188 @@
+	 * @throws ParserConfigurationException
+	 * @throws SAXException
+	 */
+	public static String getPublicationHtml(String xmlDocUrl, ServletContext context)
```
I don't see where the context is being used. Can it be removed?

orosz-usgs replied:
yes, not used anymore.

dsteinich commented on src/main/java/gov/usgs/cida/pubs/utility/XmlUtils.java:
```diff
@@ -0,0 +1,188 @@
+	 */
+	public static String getPublicationHtml(String xmlDocUrl, ServletContext context)
+			throws TransformerException, IOException, ParserConfigurationException, SAXException {
+		File xsltDir = new ClassPathResource(xsltResourceDir).getFile();
```
Could this be simpler by just going after the style sheet in one step? All of these files are part of the jar, so we really should never have a not found situation?

orosz-usgs replied:
When I tried getting the stylesheet directory as a class path resource it did not find it.
Because I think, the xslt directory is in the class path, but not the contents.

dsteinich commented on src/main/java/gov/usgs/cida/pubs/utility/XmlUtils.java:
```diff
@@ -0,0 +1,188 @@
+
+	private static final List<String> IMAGE_NODE_NAMES = List.of("graphic");
+
+	public static final String xsltResourceDir = "xslt";
```
Should these two fields have upper case & underscored names?

dsteinich commented on src/main/java/gov/usgs/cida/pubs/utility/XmlUtils.java:
```diff
@@ -0,0 +1,188 @@
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+
+public class XmlUtils {
```
Curious on why this is a utils class with static methods, rather than a service?

dsteinich commented on src/main/java/gov/usgs/cida/pubs/utility/XmlUtils.java:
```diff
@@ -0,0 +1,188 @@
+	public static final String pubsStyleSheet = "pubs-html.xsl";
+
+	// TODO: make this a configurable parameter
+	// This is a url for pulling images from SPN
```
If you make this a service class, use the appropriate spring annotation, and inject it into the PwPublicationMvcService, then you can inject the ConfigurationService into it and take care of this TODO.

dsteinich commented on src/main/java/gov/usgs/cida/pubs/webservice/pw/PwPublicationMvcService.java:
```diff
@@ -214,6 +220,46 @@ public void getBulkCrossref(HttpServletRequest request,
+	)
+	public void getPublicationHtml(HttpServletRequest request,
+						HttpServletResponse response, @PathVariable("indexId") String indexId) throws IOException {
+		PwPublication publication = getPwPublicationJSON(indexId, response);
```
Please call busService.getByIndexId(indexId) directly

dsteinich commented on src/main/java/gov/usgs/cida/pubs/webservice/pw/PwPublicationMvcService.java:
```diff
@@ -214,6 +220,46 @@ public void getBulkCrossref(HttpServletRequest request,
+			try {
+				htmlOutput = XmlUtils.getPublicationHtml(xmlUrl, request.getServletContext());
+			} catch (Exception ex) {
+				response.setStatus(HttpStatus.CONFLICT.value());
```
We typically use a HttpStatus.INTERNAL_SERVER_ERROR.value() for unhandled/unexpected exceptions and let the GlobalDefaultExceptionHandler catch, process, and sanitize them.

Commit:
PUBSTWO-1690 (Make transformed XML pubs available from pubs-services):
  Cleanup from review.

dsteinich commented on src/test/java/gov/usgs/cida/pubs/webservice/pw/PwPublicationMvcServiceHtmlIT.java:
```diff
@@ -0,0 +1,111 @@
+	private MvcResult doGetPublicationHtml(String indexId, ResultMatcher expectedStatus) throws Exception {
+		MockHttpServletRequestBuilder request = getPubHtmlReq("/publication/full/" + indexId);
+		MvcResult rtn = mockMvc.perform(request)
+			//.andExpect(expectedStatus)
```
Can this be reactivated or the tests modified to verify the return status?

orosz-usgs replied:
yes, put back in.

dsteinich commented on src/test/java/gov/usgs/cida/pubs/webservice/pw/PwPublicationMvcServiceHtmlIT.java:
```diff
@@ -0,0 +1,111 @@
+	@DatabaseSetup("classpath:/testData/publicationSubtype.xml"),
+	@DatabaseSetup("classpath:/testData/publicationStream.xml")
+})
+public class PwPublicationMvcServiceHtmlIT extends BaseIT {
```
Given that you've mocked the dao, I wouldn't call this an integration test. If you change the name, base class, and remove the database setup it should run faster.

Commit:
PUBSTWO-1690 (Make transformed XML pubs available from pubs-services):
  From review:
    removed unused ServletContext context parameter from XmlUtils.getPublicationHtml(xmlUrl)
    Changed define upper case & underscored name PUBS_STYLESHEET
    Only look once for the style sheet, removed check for xslt resource directory
    Converted XmlUtils to a service
    Renamed PwPublicationMvcServiceHtmlIT to PwPublicationMvcServiceHtml and took out database setup

Commit:
PUBSTWO-1690 (Make transformed XML pubs available from pubs-services):
 Added missing: src/main/java/gov/usgs/cida/pubs/busservice/intfc/IXmlBusService.java

dsteinich commented on src/main/java/gov/usgs/cida/pubs/webservice/pw/PwPublicationMvcService.java:
```diff
@@ -245,9 +249,9 @@ public void getPublicationHtml(HttpServletRequest request,
+				htmlOutput = xmlBusService.getPublicationHtml(xmlUrl);
 			} catch (Exception ex) {
-				response.setStatus(HttpStatus.CONFLICT.value());
+				response.setStatus(HttpStatus.INTERNAL_SERVER_ERROR.value());
```
The exception output is still not sanitized doing it this way.

dsteinich commented on src/test/java/gov/usgs/cida/pubs/webservice/pw/PwPublicationMvcServiceHtmlTest.java:
```diff
@@ -22,69 +23,68 @@
 			FreemarkerConfig.class, PublicationBusService.class, PwPublication.class,
-			PwPublicationDao.class, PublicationDao.class, SpringConfig.class,
-			CustomStringToArrayConverter.class, StringArrayCleansingConverter.class,
+			SpringConfig.class, CustomStringToArrayConverter.class, StringArrayCleansingConverter.class,
```
Don't do this. Bringing in SpringConfig.class loads almost the entire Spring context.
This class and it's tests should be able to be done similar to MpPublicationMvcServiceTest.

Commit:
PUBSTWO-1690 (Make transformed XML pubs available from pubs-services):
  codacy fix.

dsteinich commented on src/main/java/gov/usgs/cida/pubs/busservice/xml/XmlBusService.java:
```diff
@@ -134,13 +154,28 @@ public static String getDocumentHtml(URL xmlDoc, File xlsStylesheet, boolean val
 	}
 
-	private static void updateImageLinks(URL xmlUrl, Node node) {
+	// make sure the image link to add to the html is a valid Url
```
I don't think we should be validating the environment/configuration variables. They should be taken as is and if wrong, the application allowed to malfunction. That way any problems will hopefully be caught soon after a fresh deployment.

dsteinich reviewed (changes requested):
For extra credit, it would be nice to fork JATS, make/deploy it as a jar & just include that jar as a dependency. That should better isolate the 80 some related files, making both them and pubs-services easier to maintain...

Commit:
PUBSTWO-1690 (Make transformed XML pubs available from pubs-services):
  Removed check of configuration parameter in XmlBusService
  Removed unneeded classes from PwPublicationMvcServiceHtmlTest @SpringBootTest annotation.
  Sanitized error message in PwPublicationMvcService.getPublicationHtml

Commit:
PUBSTWO-1690 (Make transformed XML pubs available from pubs-services):
  Removed unused imports in XmlBusService

Commit:
PUBSTWO-1690 (Make transformed XML pubs available from pubs-services):
  Removed try catch in PwPublicationMvcService.getPublicationHtml in favor of global exception handler
